### PR TITLE
Added a pydantic-validated Camera class and test.

### PR DESCRIPTION
**Pull request by JuanCab**
I have added a pydantic (v1) validated Camera class.  I also had to modify previous units checks (we now through `ValueError` instead of `TypeError`) and added a new check to see that any 'count' (e.g. count or electron or whatever) is acceptable as long as the units of read_noise, gain, and dark_current are consistent.  

I also confirmed we can json serialize the Camera object, so we can dump its value to disk and should be able to deserialize it easily enough.

### CODE REVIEW

File: stellarphot/tests/test_core.py [hunk 1]
```diff
@@ -32,28 +32,44 @@ def test_camera_unitscheck():
     dark_current = 0.01 * u.electron / u.second
     pixel_scale = 0.563 * u.arcsec / u.pix
 
-    with pytest.raises(TypeError):
+    with pytest.raises(ValueError):
```
Comment (mwcraig) [inline stellarphot/tests/test_core.py:35]: I'm confused about why this passes. pydantic should have turned the `ValueError` into a `ValidationError`.  I'd suggest we check for that , and add some `match` language to make sure we are getting the expected failure:

```suggestion
    with pytest.raises(ValidationError,  match="gain"):
```

File: stellarphot/tests/test_core.py [hunk 2]
```diff
@@ -32,28 +32,44 @@ def test_camera_unitscheck():
...
         c = Camera(gain=gain.value,
                 read_noise=read_noise,
                 dark_current=dark_current,
                 pixel_scale=pixel_scale)
-    with pytest.raises(TypeError):
+    with pytest.raises(ValueError):
```
Comment (mwcraig) [inline stellarphot/tests/test_core.py:40]: As above

File: stellarphot/tests/test_core.py [hunk 3]
```diff
@@ -32,28 +32,44 @@ def test_camera_unitscheck():
...
         c = Camera(gain=gain,
                 read_noise=read_noise.value,
                 dark_current=dark_current,
                 pixel_scale=pixel_scale)
-    with pytest.raises(TypeError):
+    with pytest.raises(ValueError):
```
Comment (mwcraig) [inline stellarphot/tests/test_core.py:45]: Ditto

File: stellarphot/tests/test_core.py [hunk 4]
```diff
@@ -32,28 +32,44 @@ def test_camera_unitscheck():
...
         c = Camera(gain=gain,
                 read_noise=read_noise,
                 dark_current=dark_current.value,
                 pixel_scale=pixel_scale)
-    with pytest.raises(TypeError):
+    with pytest.raises(ValueError):
```
Comment (mwcraig) [inline stellarphot/tests/test_core.py:50]: one more time!

File: stellarphot/core.py [hunk 5]
```diff
@@ -70,31 +111,50 @@ class Camera:
...
+        json_encoders = {
+            Quantity: lambda v: f"{v.value} {v.unit}",
+            QuantityType: lambda v: f"{v.value} {v.unit}",
+            PixelScaleType: lambda v: f"{v.value} {v.unit}"
+        }
+
+    # When the switch to pydantic v2 happens, this root_validator will need
+    # to be replaced by a model_validator decorator.
+    @root_validator()
+    def validate_gain(cls, values):
```
Comment (mwcraig) [inline stellarphot/core.py:132]: Also, `root_validator` is a classmethod accoring to the documentation, so let's mark it that way:

```suggestion
    @classmethod
    def validate_gain(cls, values):
```

**Review by mwcraig — CHANGES_REQUESTED** [review]
Despite the extensive comments I think this is really close to ready. Sorry it took a while -- I needed some uninterrupted time so I could try things locally since I'm not that familiar with pydantic or Quantity.

In addition to the line suggestions, I have one other: 

Add an "image_unit" attribute to `Camera` so that the user has the option of using whatever image unit that want (ADU, count, something that is already gain correct so that it is in electrons, say)

What do you think?

File: stellarphot/core.py [hunk 6]
```diff
@@ -2,30 +2,71 @@
...
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
+        try:
+            v = Quantity(v)
+        except TypeError:
+            raise ValueError(f"Invalid value for Quantity: {v}")
```
[the remark this answers is not in the document]
Comment (JuanCab) [reply stellarphot/core.py:30]: Good catch, I forgot unitless is allowed.  I've implemented your fix.

File: stellarphot/core.py [hunk 7]
```diff
@@ -2,30 +2,71 @@
...
```
[the remark this answers is not in the document]
Comment (JuanCab) [reply stellarphot/core.py:46]: Yeah, that's why... at least I assumed it operated that way with all validate functions getting run down the class heirarchy.

File: stellarphot/core.py [hunk 8]
```diff
@@ -70,31 +111,50 @@ class Camera:
...
+            QuantityType: lambda v: f"{v.value} {v.unit}",
+            PixelScaleType: lambda v: f"{v.value} {v.unit}"
+        }
+
+    # When the switch to pydantic v2 happens, this root_validator will need
+    # to be replaced by a model_validator decorator.
+    @root_validator()
+    def validate_gain(cls, values):
+        # Get read noise units
+        try:
```
[the remark this answers is not in the document]
Comment (JuanCab) [reply stellarphot/core.py:134]: Yeah, I think this is a left over from debugging the code and my trying to figure out what was happening.  Fixing it.

File: stellarphot/core.py [hunk 9]
```diff
@@ -70,31 +111,50 @@ class Camera:
...
+            rn_unit = Quantity(values['read_noise']).unit
+        except KeyError:
+            raise ValueError("read_noise must be specified including units.")
+
+        # Check that gain and read noise have compatible units, that is that
+        # gain is read noise per adu.
+        gain = Quantity(values['gain'])
```
[the remark this answers is not in the document]
Comment (JuanCab) [reply stellarphot/core.py:141]: Again, left over from writing the code, good catch.

File: stellarphot/tests/test_core.py (hunk 1 above)
Comment (JuanCab) [reply stellarphot/tests/test_core.py:35]: Well, we are throwing `ValueError` in the Camera class if there are bad values, should I have been throwing `ValidationError`?

File: stellarphot/tests/test_core.py (hunk 2 above)
Comment (JuanCab) [reply stellarphot/tests/test_core.py:40]: Noted

File: stellarphot/tests/test_core.py (hunk 4 above)
Comment (JuanCab) [reply stellarphot/tests/test_core.py:50]: Noted.

File: stellarphot/core.py (hunk 5 above)
Comment (JuanCab) [reply stellarphot/core.py:132]: [The documentation I looked at](https://docs.pydantic.dev/1.10/usage/validators/) never shows `@classmethod` added in addition to `@root_validator`, can you point mt to your documentation reference?

File: stellarphot/core.py [hunk 10]
```diff
@@ -70,31 +111,50 @@ class Camera:
...
+        extra = "forbid"
+        json_encoders = {
+            Quantity: lambda v: f"{v.value} {v.unit}",
+            QuantityType: lambda v: f"{v.value} {v.unit}",
+            PixelScaleType: lambda v: f"{v.value} {v.unit}"
+        }
+
+    # When the switch to pydantic v2 happens, this root_validator will need
+    # to be replaced by a model_validator decorator.
+    @root_validator()
```
[the remark this answers is not in the document]
Comment (JuanCab) [reply stellarphot/core.py:131]: Addressed.

File: stellarphot/tests/test_core.py (hunk 1 above)
Comment (JuanCab) [reply stellarphot/tests/test_core.py]: OK, I changed the tests, and they still clear.

File: stellarphot/core.py (hunk 5 above)
Comment (JuanCab) [reply stellarphot/core.py:135]: OK, you win, changes made.

**Review by mwcraig — APPROVED** [review]
Thanks!

**Pull request merged**